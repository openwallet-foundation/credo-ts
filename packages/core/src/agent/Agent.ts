import type { AgentDependencies } from './AgentDependencies'
import type { AgentModulesInput } from './AgentModules'
import type { AgentMessageReceivedEvent } from './Events'
import type { Module } from '../plugins'
import type { InboundTransport } from '../transport/InboundTransport'
import type { OutboundTransport } from '../transport/OutboundTransport'
import type { InitConfig } from '../types'
import type { Subscription } from 'rxjs'

import { Subject } from 'rxjs'
import { concatMap, takeUntil } from 'rxjs/operators'

import { InjectionSymbols } from '../constants'
import { SigningProviderToken } from '../crypto'
import { JwsService } from '../crypto/JwsService'
import { CredoError } from '../error'
import { DependencyManager } from '../plugins'
import { DidCommMessageRepository, StorageUpdateService, StorageVersionRepository } from '../storage'

import { AgentConfig } from './AgentConfig'
import { extendModulesWithDefaultModules } from './AgentModules'
import { BaseAgent } from './BaseAgent'
import { Dispatcher } from './Dispatcher'
import { EnvelopeService } from './EnvelopeService'
import { EventEmitter } from './EventEmitter'
import { AgentEventTypes } from './Events'
import { FeatureRegistry } from './FeatureRegistry'
import { MessageHandlerRegistry } from './MessageHandlerRegistry'
import { MessageReceiver } from './MessageReceiver'
import { MessageSender } from './MessageSender'
import { TransportService } from './TransportService'
import { AgentContext, DefaultAgentContextProvider } from './context'

interface AgentOptions<AgentModules extends AgentModulesInput> {
  config: InitConfig
  modules?: AgentModules
  dependencies: AgentDependencies
}

// Any makes sure you can use Agent as a type without always needing to specify the exact generics for the agent
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class Agent<AgentModules extends AgentModulesInput = any> extends BaseAgent<AgentModules> {
  private messageSubscription?: Subscription

  public constructor(options: AgentOptions<AgentModules>, dependencyManager = new DependencyManager()) {
    const agentConfig = new AgentConfig(options.config, options.dependencies)
    const modulesWithDefaultModules = extendModulesWithDefaultModules(options.modules)

    // Register internal dependencies
    dependencyManager.registerSingleton(MessageHandlerRegistry)
    dependencyManager.registerSingleton(EventEmitter)
    dependencyManager.registerSingleton(MessageSender)
    dependencyManager.registerSingleton(MessageReceiver)
    dependencyManager.registerSingleton(TransportService)
    dependencyManager.registerSingleton(Dispatcher)
    dependencyManager.registerSingleton(EnvelopeService)
    dependencyManager.registerSingleton(FeatureRegistry)
    dependencyManager.registerSingleton(JwsService)
    dependencyManager.registerSingleton(DidCommMessageRepository)
    dependencyManager.registerSingleton(StorageVersionRepository)
    dependencyManager.registerSingleton(StorageUpdateService)

    // This is a really ugly hack to make tsyringe work without any SigningProviders registered
    // It is currently impossible to use @injectAll if there are no instances registered for the
    // token. We register a value of `default` by default and will filter that out in the registry.
    // Once we have a signing provider that should always be registered we can remove this. We can make an ed25519
    // signer using the @stablelib/ed25519 library.
    dependencyManager.registerInstance(SigningProviderToken, 'default')

    dependencyManager.registerInstance(AgentConfig, agentConfig)
    dependencyManager.registerInstance(InjectionSymbols.AgentDependencies, agentConfig.agentDependencies)
    dependencyManager.registerInstance(InjectionSymbols.Stop$, new Subject<boolean>())
    dependencyManager.registerInstance(InjectionSymbols.FileSystem, new agentConfig.agentDependencies.FileSystem())

    // Register all modules. This will also include the default modules
    dependencyManager.registerModules(modulesWithDefaultModules)

    // Register possibly already defined services
    if (!dependencyManager.isRegistered(InjectionSymbols.Wallet)) {
      throw new CredoError(
        "Missing required dependency: 'Wallet'. You can register it using the AskarModule, or implement your own."
      )
    }
    if (!dependencyManager.isRegistered(InjectionSymbols.Logger)) {
      dependencyManager.registerInstance(InjectionSymbols.Logger, agentConfig.logger)
    }
    if (!dependencyManager.isRegistered(InjectionSymbols.StorageService)) {
      throw new CredoError(
        "Missing required dependency: 'StorageService'. You can register it using the AskarModule, or implement your own."
      )
    }

    // TODO: contextCorrelationId for base wallet
    // Bind the default agent context to the container for use in modules etc.
    dependencyManager.registerInstance(
      AgentContext,
      new AgentContext({
        dependencyManager,
        contextCorrelationId: 'default',
      })
    )

    // If no agent context provider has been registered we use the default agent context provider.
    if (!dependencyManager.isRegistered(InjectionSymbols.AgentContextProvider)) {
      dependencyManager.registerSingleton(InjectionSymbols.AgentContextProvider, DefaultAgentContextProvider)
    }

    super(agentConfig, dependencyManager)
  }

  public registerInboundTransport(inboundTransport: InboundTransport) {
    this.messageReceiver.registerInboundTransport(inboundTransport)
  }

  public async unregisterInboundTransport(inboundTransport: InboundTransport) {
    await this.messageReceiver.unregisterInboundTransport(inboundTransport)
  }

  public get inboundTransports() {
    return this.messageReceiver.inboundTransports
  }

  public registerOutboundTransport(outboundTransport: OutboundTransport) {
    this.messageSender.registerOutboundTransport(outboundTransport)
  }

  public async unregisterOutboundTransport(outboundTransport: OutboundTransport) {
    await this.messageSender.unregisterOutboundTransport(outboundTransport)
  }

  public get outboundTransports() {
    return this.messageSender.outboundTransports
  }

  public get events() {
    return this.eventEmitter
  }

  /**
   * Agent's feature registry
   */
  public get features() {
    return this.featureRegistry
  }

  public async initialize() {
    const stop$ = this.dependencyManager.resolve<Subject<boolean>>(InjectionSymbols.Stop$)

    // Listen for new messages (either from transports or somewhere else in the framework / extensions)
    // We create this before doing any other initialization, so the initialization could already receive messages
    this.messageSubscription = this.eventEmitter
      .observable<AgentMessageReceivedEvent>(AgentEventTypes.AgentMessageReceived)
      .pipe(
        takeUntil(stop$),
        concatMap((e) =>
          this.messageReceiver
            .receiveMessage(e.payload.message, {
              connection: e.payload.connection,
              contextCorrelationId: e.payload.contextCorrelationId,
              receivedAt: e.payload.receivedAt,
            })
            .catch((error) => {
              this.logger.error('Failed to process message', { error })
            })
        )
      )
      .subscribe()

    await super.initialize()

    for (const [, module] of Object.entries(this.dependencyManager.registeredModules) as [string, Module][]) {
      if (module.initialize) {
        await module.initialize(this.agentContext)
      }
    }

    for (const transport of this.inboundTransports) {
      await transport.start(this)
    }

    for (const transport of this.outboundTransports) {
      await transport.start(this)
    }

    // Connect to mediator through provided invitation if provided in config
    // Also requests mediation ans sets as default mediator
    // Because this requires the connections module, we do this in the agent constructor
    if (this.mediationRecipient.config.mediatorInvitationUrl) {
      this.logger.debug('Provision mediation with invitation', {
        mediatorInvitationUrl: this.mediationRecipient.config.mediatorInvitationUrl,
      })
      const mediationConnection = await this.getMediationConnection(
        this.mediationRecipient.config.mediatorInvitationUrl
      )
      await this.mediationRecipient.provision(mediationConnection)
    }

    await this.messagePickup.initialize()
    await this.mediator.initialize()
    await this.mediationRecipient.initialize()

    this._isInitialized = true
  }

  public async shutdown() {
    const stop$ = this.dependencyManager.resolve<Subject<boolean>>(InjectionSymbols.Stop$)
    // All observables use takeUntil with the stop$ observable
    // this means all observables will stop running if a value is emitted on this observable
    stop$.next(true)

    // Stop transports
    const allTransports = [...this.inboundTransports, ...this.outboundTransports]
    const transportPromises = allTransports.map((transport) => transport.stop())
    await Promise.all(transportPromises)

    if (this.wallet.isInitialized) {
      await this.wallet.close()
    }

    this._isInitialized = false
  }

  protected async getMediationConnection(mediatorInvitationUrl: string) {
    const outOfBandInvitation = await this.oob.parseInvitation(mediatorInvitationUrl)
    const outOfBandRecord = await this.oob.findByReceivedInvitationId(outOfBandInvitation.id)
    const [connection] = outOfBandRecord ? await this.connections.findAllByOutOfBandId(outOfBandRecord.id) : []

    if (!connection) {
      this.logger.debug('Mediation connection does not exist, creating connection')
      // We don't want to use the current default mediator when connecting to another mediator
      const routing = await this.mediationRecipient.getRouting({ useDefaultMediator: false })

      this.logger.debug('Routing created', routing)
      const { connectionRecord: newConnection } = await this.oob.receiveInvitation(outOfBandInvitation, {
        routing,
      })
      this.logger.debug(`Mediation invitation processed`, { outOfBandInvitation })

      if (!newConnection) {
        throw new CredoError('No connection record to provision mediation.')
      }

      return this.connections.returnWhenIsConnected(newConnection.id)
    }

    if (!connection.isReady) {
      return this.connections.returnWhenIsConnected(connection.id)
    }
    return connection
  }
}
