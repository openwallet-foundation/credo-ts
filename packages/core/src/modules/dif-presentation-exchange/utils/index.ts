export * from './transform'
export * from './credentialSelection'
export * from './presentationsToCreate'
export * from './presentationSelection'
