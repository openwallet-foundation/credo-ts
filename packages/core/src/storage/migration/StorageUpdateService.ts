import type { AgentContext } from '../../agent'
import type { VersionString } from '../../utils/version'
import type { UpdateToVersion } from './updates'

import { InjectionSymbols } from '../../constants'
import { Logger } from '../../logger'
import { inject, injectable } from '../../plugins'

import { isStorageUpToDate } from './isUpToDate'
import { StorageVersionRecord } from './repository/StorageVersionRecord'
import { StorageVersionRepository } from './repository/StorageVersionRepository'
import { INITIAL_STORAGE_VERSION } from './updates'

@injectable()
export class StorageUpdateService {
  private static STORAGE_VERSION_RECORD_ID = 'STORAGE_VERSION_RECORD_ID'

  private logger: Logger
  private storageVersionRepository: StorageVersionRepository

  public constructor(
    @inject(InjectionSymbols.Logger) logger: Logger,
    storageVersionRepository: StorageVersionRepository
  ) {
    this.logger = logger
    this.storageVersionRepository = storageVersionRepository
  }

  public async isUpToDate(agentContext: AgentContext, updateToVersion?: UpdateToVersion) {
    const currentStorageVersion = await this.getCurrentStorageVersion(agentContext)
    return isStorageUpToDate(currentStorageVersion, updateToVersion)
  }

  public async getCurrentStorageVersion(agentContext: AgentContext): Promise<VersionString> {
    const storageVersionRecord = await this.getStorageVersionRecord(agentContext)

    return storageVersionRecord.storageVersion
  }

  public async setCurrentStorageVersion(agentContext: AgentContext, storageVersion: VersionString) {
    this.logger.debug(`Setting current agent storage version to ${storageVersion}`)
    const storageVersionRecord = await this.storageVersionRepository.findById(
      agentContext,
      StorageUpdateService.STORAGE_VERSION_RECORD_ID
    )

    if (!storageVersionRecord) {
      this.logger.trace('Storage upgrade record does not exist yet. Creating.')
      await this.storageVersionRepository.save(
        agentContext,
        new StorageVersionRecord({
          id: StorageUpdateService.STORAGE_VERSION_RECORD_ID,
          storageVersion,
        })
      )
    } else {
      this.logger.trace('Storage upgrade record already exists. Updating.')
      storageVersionRecord.storageVersion = storageVersion
      await this.storageVersionRepository.update(agentContext, storageVersionRecord)
    }
  }

  /**
   * Retrieve the update record, creating it if it doesn't exist already.
   *
   * The storageVersion will be set to the INITIAL_STORAGE_VERSION if it doesn't exist yet,
   * as we can assume the wallet was created before the update record existed
   */
  public async getStorageVersionRecord(agentContext: AgentContext) {
    let storageVersionRecord = await this.storageVersionRepository.findById(
      agentContext,
      StorageUpdateService.STORAGE_VERSION_RECORD_ID
    )

    if (!storageVersionRecord) {
      storageVersionRecord = new StorageVersionRecord({
        id: StorageUpdateService.STORAGE_VERSION_RECORD_ID,
        storageVersion: INITIAL_STORAGE_VERSION,
      })
      await this.storageVersionRepository.save(agentContext, storageVersionRecord)
    }

    return storageVersionRecord
  }
}
