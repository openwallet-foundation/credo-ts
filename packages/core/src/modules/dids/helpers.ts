import { Key, KeyType } from '../../crypto'
import { isDid } from '../../utils'

import { DidKey } from './methods/key'

export function isDidKey(key: string) {
  return isDid(key, 'key')
}

export function didKeyToVerkey(key: string) {
  if (isDidKey(key)) {
    const publicKeyBase58 = DidKey.fromDid(key).key.publicKeyBase58
    return publicKeyBase58
  }
  return key
}

export function verkeyToDidKey(key: string) {
  if (isDidKey(key)) return key
  const publicKeyBase58 = key
  const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
  const didKey = new DidKey(ed25519Key)
  return didKey.did
}

export function didKeyToInstanceOfKey(key: string) {
  const didKey = DidKey.fromDid(key)
  return didKey.key
}

export function verkeyToInstanceOfKey(verkey: string) {
  const ed25519Key = Key.fromPublicKeyBase58(verkey, KeyType.Ed25519)
  return ed25519Key
}
