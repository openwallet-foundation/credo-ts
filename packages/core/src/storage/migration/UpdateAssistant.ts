import type { BaseAgent } from '../../agent/BaseAgent'
import type { Module } from '../../plugins'
import type { Update, UpdateConfig, UpdateToVersion } from './updates'

import { CredoError } from '../../error'
import { isFirstVersionEqualToSecond, isFirstVersionHigherThanSecond, parseVersionString } from '../../utils/version'

import { StorageUpdateService } from './StorageUpdateService'
import { StorageUpdateError } from './error/StorageUpdateError'
import { CURRENT_FRAMEWORK_STORAGE_VERSION, DEFAULT_UPDATE_CONFIG, supportedUpdates } from './updates'

export interface UpdateAssistantUpdateOptions {
  updateToVersion?: UpdateToVersion
}

// biome-ignore lint/suspicious/noExplicitAny: <explanation>
export class UpdateAssistant<Agent extends BaseAgent<any> = BaseAgent> {
  private agent: Agent
  private storageUpdateService: StorageUpdateService
  private updateConfig: UpdateConfig

  public constructor(agent: Agent, updateConfig: UpdateConfig = DEFAULT_UPDATE_CONFIG) {
    this.agent = agent
    this.updateConfig = updateConfig

    this.storageUpdateService = this.agent.dependencyManager.resolve(StorageUpdateService)
  }

  public async initialize() {
    if (this.agent.isInitialized) {
      throw new CredoError("Can't initialize UpdateAssistant after agent is initialized")
    }
  }

  public async isUpToDate(updateToVersion?: UpdateToVersion) {
    return this.storageUpdateService.isUpToDate(this.agent.context, updateToVersion)
  }

  public async getCurrentAgentStorageVersion() {
    return this.storageUpdateService.getCurrentStorageVersion(this.agent.context)
  }

  public static get frameworkStorageVersion() {
    return CURRENT_FRAMEWORK_STORAGE_VERSION
  }

  public async getNeededUpdates(toVersion?: UpdateToVersion) {
    const currentStorageVersion = parseVersionString(
      await this.storageUpdateService.getCurrentStorageVersion(this.agent.context)
    )

    const parsedToVersion = toVersion ? parseVersionString(toVersion) : undefined

    // If the current storage version is higher or equal to the toVersion, we can't update, so return empty array
    if (
      parsedToVersion &&
      (isFirstVersionHigherThanSecond(currentStorageVersion, parsedToVersion) ||
        isFirstVersionEqualToSecond(currentStorageVersion, parsedToVersion))
    ) {
      return []
    }

    // Filter updates. We don't want older updates we already applied
    // or aren't needed because the wallet was created after the update script was made
    const neededUpdates = supportedUpdates.filter((update) => {
      const updateToVersion = parseVersionString(update.toVersion)

      // If the update toVersion is higher than the wanted toVersion, we skip the update
      if (parsedToVersion && isFirstVersionHigherThanSecond(updateToVersion, parsedToVersion)) {
        return false
      }

      // if an update toVersion is higher than currentStorageVersion we want to to include the update
      return isFirstVersionHigherThanSecond(updateToVersion, currentStorageVersion)
    })

    // The current storage version is too old to update
    if (
      neededUpdates.length > 0 &&
      isFirstVersionHigherThanSecond(parseVersionString(neededUpdates[0].fromVersion), currentStorageVersion)
    ) {
      throw new CredoError(
        `First fromVersion is higher than current storage version. You need to use an older version of the framework to update to at least version ${neededUpdates[0].fromVersion}`
      )
    }

    const lastUpdateToVersion = neededUpdates.length > 0 ? neededUpdates[neededUpdates.length - 1].toVersion : undefined
    if (toVersion && lastUpdateToVersion && lastUpdateToVersion !== toVersion) {
      throw new CredoError(
        `No update found for toVersion ${toVersion}. Make sure the toVersion is a valid version you can update to`
      )
    }

    return neededUpdates
  }

  public async update(options?: UpdateAssistantUpdateOptions) {
    const updateIdentifier = Date.now().toString()
    const updateToVersion = options?.updateToVersion

    try {
      this.agent.config.logger.info(`Starting update of agent storage with updateIdentifier ${updateIdentifier}`)
      const neededUpdates = await this.getNeededUpdates(updateToVersion)

      const currentStorageVersion = parseVersionString(
        await this.storageUpdateService.getCurrentStorageVersion(this.agent.context)
      )
      const parsedToVersion = updateToVersion ? parseVersionString(updateToVersion) : undefined

      // If the current storage version is higher or equal to the toVersion, we can't update.
      if (
        parsedToVersion &&
        (isFirstVersionHigherThanSecond(currentStorageVersion, parsedToVersion) ||
          isFirstVersionEqualToSecond(currentStorageVersion, parsedToVersion))
      ) {
        throw new StorageUpdateError(
          `Can't update to version ${updateToVersion} because it is lower or equal to the current agent storage version ${currentStorageVersion[0]}.${currentStorageVersion[1]}}`
        )
      }

      if (neededUpdates.length === 0) {
        this.agent.config.logger.info('No update needed. Agent storage is up to date.')
        return
      }

      const fromVersion = neededUpdates[0].fromVersion
      const toVersion = neededUpdates[neededUpdates.length - 1].toVersion

      this.agent.config.logger.info(
        `Starting update process. Total of ${neededUpdates.length} update(s) will be applied to update the agent storage from version ${fromVersion} to version ${toVersion}`
      )

      try {
        for (const update of neededUpdates) {
          const registeredModules = Object.values(this.agent.dependencyManager.registeredModules)
          const modulesWithUpdate: Array<{ module: Module; update: Update }> = []

          // Filter modules that have an update script for the current update
          for (const registeredModule of registeredModules) {
            const moduleUpdate = registeredModule.updates?.find(
              (module) => module.fromVersion === update.fromVersion && module.toVersion === update.toVersion
            )

            if (moduleUpdate) {
              modulesWithUpdate.push({
                module: registeredModule,
                update: moduleUpdate,
              })
            }
          }

          this.agent.config.logger.info(
            `Starting update of agent storage from version ${update.fromVersion} to version ${update.toVersion}. Found ${modulesWithUpdate.length} extension module(s) with update scripts`
          )
          await update.doUpdate(this.agent)

          this.agent.config.logger.info(
            `Finished update of core agent storage from version ${update.fromVersion} to version ${update.toVersion}. Starting update of extension modules`
          )

          for (const moduleWithUpdate of modulesWithUpdate) {
            this.agent.config.logger.info(
              `Starting update of extension module ${moduleWithUpdate.module.constructor.name} from version ${moduleWithUpdate.update.fromVersion} to version ${moduleWithUpdate.update.toVersion}`
            )
            await moduleWithUpdate.update.doUpdate(this.agent, this.updateConfig)
            this.agent.config.logger.info(
              `Finished update of extension module ${moduleWithUpdate.module.constructor.name} from version ${moduleWithUpdate.update.fromVersion} to version ${moduleWithUpdate.update.toVersion}`
            )
          }

          // Update the framework version in storage
          await this.storageUpdateService.setCurrentStorageVersion(this.agent.context, update.toVersion)
          this.agent.config.logger.info(
            `Successfully updated agent storage from version ${update.fromVersion} to version ${update.toVersion}`
          )
        }
      } catch (error) {
        this.agent.config.logger.fatal('An error occurred while updating the wallet.', {
          error,
        })

        throw error
      }
    } catch (error) {
      this.agent.config.logger.error(`Error updating storage (updateIdentifier: ${updateIdentifier})`, {
        cause: error,
      })

      throw new StorageUpdateError(`Error updating storage (updateIdentifier: ${updateIdentifier}): ${error.message}`, {
        cause: error,
      })
    }

    return updateIdentifier
  }
}
