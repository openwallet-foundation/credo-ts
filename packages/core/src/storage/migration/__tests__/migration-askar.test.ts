import { readFileSync } from 'fs'
import path from 'path'

import { CredentialExchangeRecord, CredentialRepository } from '../../../../../didcomm/src/modules/credentials'
import { getAgentOptions } from '../../../../tests/helpers'
import { Agent } from '../../../agent/Agent'
import { JsonTransformer } from '../../../utils'
import { StorageUpdateService } from '../StorageUpdateService'
import { UpdateAssistant } from '../UpdateAssistant'

const agentOptions = getAgentOptions('UpdateAssistant | Aries Askar', undefined, undefined, undefined, {
  requireDidcomm: true,
})

const aliceCredentialRecordsString = readFileSync(
  path.join(__dirname, '__fixtures__/alice-4-credentials-0.1.json'),
  'utf8'
)

const backupDate = new Date('2022-03-22T22:50:20.522Z')
jest.useFakeTimers().setSystemTime(backupDate)

describe('UpdateAssistant | Aries Askar', () => {
  let updateAssistant: UpdateAssistant
  let agent: Agent

  beforeEach(async () => {
    agent = new Agent(agentOptions)
    updateAssistant = new UpdateAssistant(agent, {
      v0_1ToV0_2: {
        mediationRoleUpdateStrategy: 'allMediator',
      },
    })

    await updateAssistant.initialize()
  })

  afterEach(async () => {
    await agent.shutdown()
  })

  it('should create a backup', async () => {
    const aliceCredentialRecordsJson = JSON.parse(aliceCredentialRecordsString)

    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    const aliceCredentialRecords = Object.values(aliceCredentialRecordsJson).map((data: any) => {
      const record = JsonTransformer.fromJSON(data.value, CredentialExchangeRecord)

      record.setTags(data.tags)
      return record
    })

    const credentialRepository = agent.dependencyManager.resolve(CredentialRepository)
    const storageUpdateService = agent.dependencyManager.resolve(StorageUpdateService)

    // Add 0.1 data and set version to 0.1
    for (const credentialRecord of aliceCredentialRecords) {
      await credentialRepository.save(agent.context, credentialRecord)
    }
    await storageUpdateService.setCurrentStorageVersion(agent.context, '0.1')

    // Expect an update is needed
    expect(await updateAssistant.isUpToDate()).toBe(false)

    // Create update
    await updateAssistant.update()

    expect(await updateAssistant.isUpToDate()).toBe(true)

    expect(
      (await credentialRepository.getAll(agent.context)).sort((a, b) => a.id.localeCompare(b.id))
    ).toMatchSnapshot()
  })
})
