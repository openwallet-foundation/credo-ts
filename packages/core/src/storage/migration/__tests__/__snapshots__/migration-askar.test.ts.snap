// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAssistant | Aries Askar should create a backup 1`] = `
[
  {
    "_tags": {
      "connectionId": "0b6de73d-b376-430f-b2b4-f6e51407bb66",
      "role": "issuer",
      "state": "done",
      "threadId": "578e73da-c3be-43d4-949b-7aadfd5a6eae",
    },
    "autoAcceptCredential": "contentApproved",
    "connectionId": "0b6de73d-b376-430f-b2b4-f6e51407bb66",
    "createdAt": "2022-03-21T22:50:20.522Z",
    "credentialAttributes": [
      {
        "mime-type": "text/plain",
        "name": "name",
        "value": "Alice",
      },
      {
        "mime-type": "text/plain",
        "name": "age",
        "value": "25",
      },
      {
        "mime-type": "text/plain",
        "name": "dateOfBirth",
        "value": "2020-01-01",
      },
    ],
    "credentials": [],
    "id": "574b2a37-1db1-4af1-a3bf-35c6cb9e1d7a",
    "metadata": {
      "_internal/indyCredential": {
        "credentialDefinitionId": "TL1EaPFCZ8Si5aUrqScBDt:3:CL:681:default",
        "schemaId": "TL1EaPFCZ8Si5aUrqScBDt:2:schema-80f7eec5-8e5a-43ca-ad4d-3274fb9361b8:1.0",
      },
    },
    "protocolVersion": "v1",
    "role": "issuer",
    "state": "done",
    "threadId": "578e73da-c3be-43d4-949b-7aadfd5a6eae",
    "updatedAt": "2022-03-22T22:50:20.522Z",
  },
  {
    "_tags": {
      "connectionId": "54b61a2c-59ae-4e63-a441-7f1286350132",
      "credentialId": "a77114e1-c812-4bff-a53c-3d5003fcc278",
      "credentialIds": [
        "a77114e1-c812-4bff-a53c-3d5003fcc278",
      ],
      "role": "holder",
      "state": "done",
      "threadId": "578e73da-c3be-43d4-949b-7aadfd5a6eae",
    },
    "autoAcceptCredential": "contentApproved",
    "connectionId": "54b61a2c-59ae-4e63-a441-7f1286350132",
    "createdAt": "2022-03-21T22:50:20.535Z",
    "credentialAttributes": [
      {
        "mime-type": "text/plain",
        "name": "name",
        "value": "Alice",
      },
      {
        "mime-type": "text/plain",
        "name": "age",
        "value": "25",
      },
      {
        "mime-type": "text/plain",
        "name": "dateOfBirth",
        "value": "2020-01-01",
      },
    ],
    "credentials": [
      {
        "credentialRecordId": "a77114e1-c812-4bff-a53c-3d5003fcc278",
        "credentialRecordType": "indy",
      },
    ],
    "id": "5f2b7bc7-edfd-47e7-a1d4-aae050df2c4a",
    "metadata": {
      "_internal/indyCredential": {
        "credentialDefinitionId": "TL1EaPFCZ8Si5aUrqScBDt:3:CL:681:default",
        "schemaId": "TL1EaPFCZ8Si5aUrqScBDt:2:schema-80f7eec5-8e5a-43ca-ad4d-3274fb9361b8:1.0",
      },
      "_internal/indyRequest": {
        "master_secret_blinding_data": {
          "v_prime": "36456944381549782028917743247126995038265466209293312755125557271456380841610111892515020379470931691048072348420844231863825225515560265358581756565441268878364665494094789024845049226122885121039335781567964878826549149370097276812152226343824116049855825405977949749345353074025294938300401262824951638782220004732873597724698990420932910079362747837952520524827009393981876443737452031919055976088763615615890946142630576421462920865811255312740184209214306243871230276622595183415487741608569800898909023830922654063814555128779494528740438076748829436757078504882332589744263200806138145494157659396691564807976032319024007464003538934",
          "vr_prime": null,
        },
        "master_secret_name": "Wallet: PopulateWallet2",
        "nonce": "373984270150786864433163",
      },
    },
    "protocolVersion": "v1",
    "role": "holder",
    "state": "done",
    "threadId": "578e73da-c3be-43d4-949b-7aadfd5a6eae",
    "updatedAt": "2022-03-22T22:50:20.522Z",
  },
  {
    "_tags": {
      "connectionId": "cd66cbf1-5721-449e-8724-f4d8dcef1bc4",
      "role": "issuer",
      "state": "done",
      "threadId": "e2c2194c-6ac6-4b27-9030-18887c79b5eb",
    },
    "autoAcceptCredential": "contentApproved",
    "connectionId": "cd66cbf1-5721-449e-8724-f4d8dcef1bc4",
    "createdAt": "2022-03-21T22:50:20.740Z",
    "credentialAttributes": [
      {
        "mime-type": "text/plain",
        "name": "name",
        "value": "Alice",
      },
      {
        "mime-type": "text/plain",
        "name": "age",
        "value": "25",
      },
      {
        "mime-type": "text/plain",
        "name": "dateOfBirth",
        "value": "2020-01-01",
      },
    ],
    "credentials": [],
    "id": "ad644d8a-48a2-4c55-b46d-7a7f1a9278c7",
    "metadata": {
      "_internal/indyCredential": {
        "credentialDefinitionId": "TL1EaPFCZ8Si5aUrqScBDt:3:CL:681:default",
        "schemaId": "TL1EaPFCZ8Si5aUrqScBDt:2:schema-80f7eec5-8e5a-43ca-ad4d-3274fb9361b8:1.0",
      },
    },
    "protocolVersion": "v1",
    "role": "issuer",
    "state": "done",
    "threadId": "e2c2194c-6ac6-4b27-9030-18887c79b5eb",
    "updatedAt": "2022-03-22T22:50:20.522Z",
  },
  {
    "_tags": {
      "connectionId": "d8f23338-9e99-469a-bd57-1c9a26c0080f",
      "credentialId": "19c1f29f-d2df-486c-b8c6-950c403fa7d9",
      "credentialIds": [
        "19c1f29f-d2df-486c-b8c6-950c403fa7d9",
      ],
      "role": "holder",
      "state": "done",
      "threadId": "e2c2194c-6ac6-4b27-9030-18887c79b5eb",
    },
    "autoAcceptCredential": "contentApproved",
    "connectionId": "d8f23338-9e99-469a-bd57-1c9a26c0080f",
    "createdAt": "2022-03-21T22:50:20.746Z",
    "credentialAttributes": [
      {
        "mime-type": "text/plain",
        "name": "name",
        "value": "Alice",
      },
      {
        "mime-type": "text/plain",
        "name": "age",
        "value": "25",
      },
      {
        "mime-type": "text/plain",
        "name": "dateOfBirth",
        "value": "2020-01-01",
      },
    ],
    "credentials": [
      {
        "credentialRecordId": "19c1f29f-d2df-486c-b8c6-950c403fa7d9",
        "credentialRecordType": "indy",
      },
    ],
    "id": "c7e0a752-7f1c-41c0-b0ae-a68c2d97ca8c",
    "metadata": {
      "_internal/indyCredential": {
        "credentialDefinitionId": "TL1EaPFCZ8Si5aUrqScBDt:3:CL:681:default",
        "schemaId": "TL1EaPFCZ8Si5aUrqScBDt:2:schema-80f7eec5-8e5a-43ca-ad4d-3274fb9361b8:1.0",
      },
      "_internal/indyRequest": {
        "master_secret_blinding_data": {
          "v_prime": "24405223168730122709164916892481085040205443709643249329100687534344659826655374235392514476392517756663433844139774514430993889493707631169979521764390851593418941181409704266182779162417466204970949168472702858363964258641437554267668466400711344128132909691514606077477555576087059339291048485225394874964325220472232903203038212033940680060605090839733163438385288769519855418153181511119637865605476043416048121313638627002888436809192752657860306784733123742838413845299796745569824223645588826964796075250758249133953560017373025169692866449286962430731916293683231375510684692358406054381559324718715654332979447698704161714028193478",
          "vr_prime": null,
        },
        "master_secret_name": "Wallet: PopulateWallet2",
        "nonce": "698370616023883730498375",
      },
    },
    "protocolVersion": "v1",
    "role": "holder",
    "state": "done",
    "threadId": "e2c2194c-6ac6-4b27-9030-18887c79b5eb",
    "updatedAt": "2022-03-22T22:50:20.522Z",
  },
]
`;
